### ISSUE TITLE
Slow install time of any sensu-plugin

### ISSUE THREAD

**Issue by runningman84** [sensu-plugins/sensu-plugin#93]
The installation of any sensu plugin takes at least 10 seconds (on AWS >20s). I wonder if this process can be improved?

This is the gem debug output:
https://gist.github.com/runningman84/33996728c61b83bc4b7c

Why does it load 6 mixlib gemspec files?
Why doesn't it cache the results? (another installation takes the same time)

The total number of gems time for a typical machine (cpu, load, memory, fs, ... checks) is about 10. This means on AWS a typical sensu gem installations takes more than 4 minutes. This is quite a long time if the application itself can be installed within seconds...

**Comment by mattyjones** [comment]
I can't speak for the others but the aws gem includes the aws-sdk which is rather cumbersome.  I will do a few installs and time it myself.

**Comment by runningman84** [comment]
@mattyjones just try simple plugins like http, nginx and so on. Even these simple plugins take so long. I haven't even tried more complex plugins like aws.

I told you about AWS because somebody told me, my internet uplink might be the problem. On AWS gem installations are faster... As you can see with on a m3.medium machine the installation takes even longer.

**Comment by runningman84** [comment]
Could you please bump the mixlib dependency to 1.5?

I assume that because of this line
```
s.add_dependency('mixlib-cli', '>= 1.1.0')
```
gem looks up all mixlib-cli packages until it finds the latest release:
```
GET https://rubygems.org/quick/Marshal.4.8/mixlib-cli-1.1.0.gemspec.rz
302 Moved Temporarily
GET https://aws-eu-cache01.rubygems.org/quick/Marshal.4.8/mixlib-cli-1.1.0.gemspec.rz
200 OK
GET https://rubygems.org/quick/Marshal.4.8/mixlib-cli-1.2.0.gemspec.rz
302 Moved Temporarily
GET https://aws-eu-cache01.rubygems.org/quick/Marshal.4.8/mixlib-cli-1.2.0.gemspec.rz
200 OK
GET https://rubygems.org/quick/Marshal.4.8/mixlib-cli-1.2.2.gemspec.rz
302 Moved Temporarily
GET https://aws-eu-cache01.rubygems.org/quick/Marshal.4.8/mixlib-cli-1.2.2.gemspec.rz
200 OK
GET https://rubygems.org/quick/Marshal.4.8/mixlib-cli-1.3.0.gemspec.rz
302 Moved Temporarily
GET https://aws-eu-cache01.rubygems.org/quick/Marshal.4.8/mixlib-cli-1.3.0.gemspec.rz
200 OK
GET https://rubygems.org/quick/Marshal.4.8/mixlib-cli-1.4.0.gemspec.rz
302 Moved Temporarily
GET https://aws-eu-cache01.rubygems.org/quick/Marshal.4.8/mixlib-cli-1.4.0.gemspec.rz
200 OK
GET https://rubygems.org/quick/Marshal.4.8/mixlib-cli-1.5.0.gemspec.rz
302 Moved Temporarily
GET https://aws-eu-cache01.rubygems.org/quick/Marshal.4.8/mixlib-cli-1.5.0.gemspec.rz
200 OK
```

**Comment by mattyjones** [comment]
@portertech any objections.  Looking at release notes I don't *think* there will be any issues but would like another set of eyes on this.

**Comment by runningman84** [comment]
@mattyjones and @portertech do you think sensu_gem chef resource could cache these files?
* latest_specs.4.8.gz
* specs.4.8.gz
Right now for each sensu plugin both files are downloaded...

**Comment by runningman84** [comment]
I use jenkins to do daily kitchen tests. My setup (sensu server, client, nginx) based on ubuntu 14.04 took about 4 minutes using my old monitor cookbook (which included all check scripts within the cookbook https://github.com/runningman84/chef-monitor):

```
       Finished verifying <default-ubuntu-1404> (0m31.32s).
-----> Destroying <default-ubuntu-1404>...
       Vagrant instance <default-ubuntu-1404> destroyed.
       Finished destroying <default-ubuntu-1404> (0m18.39s).
       Finished testing <default-ubuntu-1404> (4m1.70s).
-----> Kitchen is finished. (4m2.51s)
```

The new cookbook (which uses the sensu_gem plugins for all checks) takes more than 12 minutes to setup:

```
       Finished verifying <default-ubuntu-1404> (0m32.39s).
-----> Destroying <default-ubuntu-1404>...
       Vagrant instance <default-ubuntu-1404> destroyed.
       Finished destroying <default-ubuntu-1404> (0m18.36s).
       Finished testing <default-ubuntu-1404> (12m6.90s).
-----> Kitchen is finished. (12m7.74s)
```

This is a real problem, I cannot use them in an autoscaling group anymore. In the past my instances (using sensu client + nginx server) got configured within 3-4 minutes. Today it takes more than 10 minutes. I hesitate to build golden images using packer for every service.

**Comment by obazoud** [comment]
I have the same experience here, too slow for an autoscaling group :(

**Comment by portertech** [comment]
@mattyjones pinning versions of sensu-plugin dependencies is OK :+1: Just be sure to use `gem "foo", "0.0.0"` etc.

**Comment by runningman84** [comment]
Any news?

**Comment by mattyjones** [comment]
@portertech Can you build and push this. I updated the dep and will update all the plugins as well.
This gem will go in the pipeline, I just haven't had the resources to clean it up yet and do it right.

**Comment by obazoud** [comment]
Any news?

**Comment by obazoud** [comment]
ping

**Comment by analytically** [comment]
I think this one is fixed, @portertech?

**Comment by runningman84** [comment]
the latest changes have improved the install time:
![monitor_build_time](https://cloud.githubusercontent.com/assets/1699128/11527035/beb009d8-98de-11e5-8573-62b83bd90177.PNG)

This looks good for me. Maybe we can even improve the results if the sensu chef cookbook would use the new sensu-install cmd. The cookbook could collect all gems and call sensu-install -P plugin1,plugin2,... this should ensure that the gemsepc file is only downloaded once.

**Comment by cwjohnston** [comment]
Closing as this looks like it was fixed in commit aa59019a584eae88f3e784d7079f59a762879418 which shipped in v1.3.0. 😋
